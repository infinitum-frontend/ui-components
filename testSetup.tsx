import { expect, beforeAll } from 'vitest'
import { render, RenderOptions } from '@testing-library/react'
// import '@testing-library/jest-dom/vitest' for vitest 1.x.x
import { ReactElement, JSXElementConstructor } from 'react'
import matchers from '@testing-library/jest-dom/matchers'

// extends Vitest's expect method with methods from react-testing-library
// https://github.com/testing-library/jest-dom#custom-matchers
expect.extend(matchers)

export const renderComponent = function (
  Component: ReactElement,
  options?: RenderOptions
): {
  el: HTMLElement
  container: HTMLElement
  rerender: (ui: ReactElement<any, string | JSXElementConstructor<any>>) => void
} {
  const { container, rerender } = render(Component, options)

  const el = container.firstChild as HTMLElement

  return { el, container, rerender }
}

beforeAll(() => {
  global.ResizeObserver = class ResizeObserver {
    observe(): any {}
    unobserve(): any {}
    disconnect(): any {}
  }

  // https://github.com/testing-library/react-testing-library/issues/95#issuecomment-392879866
  // https://github.com/vuejs/vue-test-utils/issues/319
  if (!HTMLElement.prototype.scrollTo) {
    HTMLElement.prototype.scrollTo = () => {}
  }
})

// runs a cleanup after each test case (e.g. clearing jsdom)
// afterEach(() => {
//   cleanup()
// })
