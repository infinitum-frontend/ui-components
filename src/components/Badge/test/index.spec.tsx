import { describe, it } from 'vitest'
import { renderComponent } from '../../../../testSetup'
import { Badge } from '../index'
import { screen } from '@testing-library/dom'

describe('badge', () => {
  it('should render', () => {
    const { el } = renderComponent(<Badge badgeContent={5} />)
    expect(el).toBeInTheDocument()
    expect(screen.getByText('5')).toBeDefined()
  })

  it('should match snapshot', () => {
    const { el } = renderComponent(<Badge badgeContent={5} />)
    expect(el).toMatchSnapshot()
  })

  it('should have position absolute if children exist', () => {
    renderComponent(<Badge badgeContent={5}>Уведомления</Badge>)
    expect(screen.getByTitle('5')).toHaveStyle('position: absolute')
  })

  it('should have position static id children don`t exist', () => {
    renderComponent(<Badge badgeContent={5} />)
    expect(screen.getByTitle('5')).toHaveStyle('position: static')
  })

  it('should support tone', () => {
    renderComponent(<Badge badgeContent={5} tone={'secondary'} />)
    expect(screen.getByTitle('5')).toHaveClass('inf-badge-sup--tone-secondary')
  })

  it('should support dot variant', () => {
    const { el } = renderComponent(<Badge dot={true} badgeContent={5} />)
    const sup = el.querySelector('sup') as HTMLElement
    expect(sup.innerHTML).toBeFalsy()
    expect(sup).toHaveStyle('border:radius: 50%')
  })

  it('should not exist if value is zero integer', () => {
    renderComponent(<Badge badgeContent={0} />)
    expect(screen.queryByTitle('0')).toBeNull()
  })

  it('should support showZero', () => {
    renderComponent(<Badge badgeContent={0} showZero={true} />)
    const el = screen.getByTitle('0')
    expect(el).toBeInTheDocument()
    expect(el.innerHTML).toBe('0')
  })

  it('should support offset', () => {
    renderComponent(
      <Badge badgeContent={5} offset={[10, 10]}>
        Уведомления
      </Badge>
    )
    expect(screen.getByTitle('5')).toHaveStyle({
      top: '10px',
      right: '-10px'
    })
  })
})
