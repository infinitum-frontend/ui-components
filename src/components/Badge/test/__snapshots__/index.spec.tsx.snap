// Vitest Snapshot v1

exports[`badge > should match snapshot 1`] = `
<span
  class="inf-badge"
>
  <sup
    class="inf-badge-sup inf-badge-sup--tone-primary inf-badge-sup--standalone"
    title="5"
  >
    5
  </sup>
</span>
`;
